### ISSUE TITLE
[flutter_tools] The tool should prefer the JDK vendored with Android Studio

### ISSUE THREAD

**Issue by AmirGamilDev** [flutter/flutter#61197]
As seen in issue #56572, if a version of JDK greater than 8 is set as `JAVA_HOME`, the tool will fail to detect android licenses. Since Android Studio vendors JDK 8, we should prefer that version.

**Comment by iapicca** [comment]
Hi @AmirGamilDev 
I'm not sure that this is exactly the case; 
keeping in mind that [android set up](https://flutter.dev/docs/get-started/install/windows#android-setup) goes through android studio inheriting openjdk8,
it is possible to use a different jdk, please see this [SO answer](https://stackoverflow.com/a/55982976).

Recommending one version is different from not supporting another one.

It is true that in some users have experienced issue with android licence
and, since isn't a procedure advised by the official docs,
imho those difficulties shouldn't fall under flutter repository scope;
nonetheless the same issue you shared offers a [solution](https://github.com/flutter/flutter/issues/56572#issuecomment-638329948) for the situation you described

Does this solve the issue for you?
Thank you

**Comment by AmirGamilDev** [comment]
Hi @iapicca 

While it's true that Android Studio itself comes bundled with Open JDK 8, a developer may _already_ (or in the future) install a more recent version of the JDK.  I don't believe it's correct to assume that a developer will be using JDK 8.

In fact, this is what happened to me.  I _did_ have JDK 8 running on my machine.  _Then_ I upgraded to JDK 14 to get around another entirely separate issue with another development environment.  _Later_, running flutter doctor gave the error message.  After much online searching, trial and error, trying solutions around unchecking hide obsolete packages, uninstalling and re-installing Android SDK tools etc. etc., I was still no further along hours later.  Eventually, I saw mention of JDK versions and because I had recently upgraded Java, I thought to try this.

If part of the Flutter philosophy is to reduce barrier to entry then I believe highlighting categorically that a developer may have issues if they don't have JDK 8 is extremely important.  To be honest, _even if_ the Flutter philosophy is not to reduce barrier to entry, out of respect for developers' time and sanity, I believe it's reasonable to expect that flutter doctor would flag this as an issue.  I don't believe developers should have to search online for hours to eventually find out that they either:

a) Have to use JDK 8 when using flutter
or
b) Have to download deprecated Java libraries and add them to the classpath in order to be able to use a more recent version of the JDK

One developer doing this is bad enough.  Countless developers doing this is worse.  And it's clear to see that a number of developers have come up against this issue.  If it's a developer's first experience of Flutter and they can't get it working after an hour or so of trying, they may just give up.  Is this what the Flutter team wants?

> Recommending one version is different from not supporting another one.

I must respectfully disagree with this.  Clearly, flutter doctor **does not work** if JDK is > 8 out of the box.  This means it's not supported.  But there is no mention of this anywhere in the docs or in flutter doctor's output.

While it's true that the Android tools SDK could probably fix this since flutter is dependant on it, I don't see the harm in flutter doctor doing the right thing and at least _warning_ a developer that they may run into issues since they are using JDK > 8.

What do you think?

**Comment by christopherfujino** [comment]
Can we not update the tool to prefer the version vendored with Android Studio (which is an android development dependency) @jonahwilliams?

**Comment by jonahwilliams** [comment]
We should if it isn't already the case, I'm not quite sure though - that code is a bit of a mess :(

**Comment by christopherfujino** [comment]
I'm going to update the action point on this issue to prefer version 8 vendored with android studio.

**Comment by AmirGamilDev** [comment]
@christopherfujino 
I'm not sure if preferring the JDK vendored with Android Studio is the way to go?  Perhaps just checking if the JDK version is greater than 8?  Or at least doesn't include the necessary classes to perform the Android licensing check.

**Comment by sebastianhaberey** [comment]
I am also not sure that preferring the JDK vendored with Android Studio is the way to go. Developers might prefer to use the same JDK on command line and Android Studio. 

For example, the integration_test package [mentions to run gradle on the command line for Android device testing](https://github.com/flutter/flutter/tree/master/packages/integration_test#android-device-testing). To have command line gradle execution and Android studio gradle execution run on the same JDK, I'd have to set my JAVA_HOME to Android Studio's internal JDK which doesn't seem like good practice.

Rather, I'd like to go the opposite way and make flutter use my standard JDK. There are a few tickets on Stackoverflow (e.g. [#1](https://stackoverflow.com/questions/66853635/how-to-change-java-version-for-flutter-package), [#2](https://stackoverflow.com/questions/66853635/how-to-change-java-version-for-flutter-package)) that indicate others have tried to go the same way.

**Comment by sebastianhaberey** [comment]
Also, [here](https://github.com/flutter/flutter/issues/81974) is an issue showing exactly what I meant: a case where `.gradlew build` worked (because it used the system JDK) while `flutter run` didnt (because it used the Android Studio JDK).

**Comment by martin-braun** [comment]
I think this issue was addressed, and I hate it. `flutter -v doctor` will always show the Java in the embedded Android Studio resources, but in my case I have issues with the latest Java 17 that ships with Jellyfish. It can't launch. 

I installed a newer Java 17 with `sdkman-cli` and now the build works in Android Studio, but Flutter will not respect my decision I made in Android Studio and it will also not respect `JAVA_HOME`.

I'm on MacOS, btw.

**Comment by gmackall** [comment]
This is an old issue, and I believe the tool currently works this way. iirc @andrewkolos you were the last one to make major changes to the handling of java versions within the flutter tool, is that correct and can you confirm?

If so feel free to close this

**Comment by andrewkolos** [comment]
> This is an old issue, and I believe the tool currently works this way. iirc [@andrewkolos[<img alt="" width="15" height="15" src="chrome-extension://dlebflppeeemcdpidccbiblndppbmjmh/ospo-chrome-ext-logo.png">](https://teams.googleplex.com/andrewkolos@google.com)](https://github.com/andrewkolos) you were the last one to make major changes to the handling of java versions within the flutter tool, is that correct and can you confirm?
> 
> If so feel free to close this

Yep, the only thing that preempts using the Android Studio-bundled JDK is one configured via `flutter config jdk-dir`.
https://github.com/flutter/flutter/blob/1709c884aaeb70f2bcbed2887d2738f30f337b93/packages/flutter_tools/lib/src/android/java.dart#L226-L247
